src/sync/semaphore.rs

```diff
@@ -555,6 +557,279 @@ mod if_alloc {
+            // GenericSharedSemaphoreAcquireFuture is stable,
+            // and we don't move any fields inside the future until it gets dropped.
+            let mut_self: &mut GenericSharedSemaphoreAcquireFuture<MutexType> =
+                unsafe { Pin::get_unchecked_mut(self) };
+
+            let semaphore = mut_self.semaphore.take().expect(
+                "polled GenericSharedSemaphoreAcquireFuture after completion",
+            );
+
+            let poll_res = unsafe {
+                let mut semaphore_state = semaphore.state.lock();
+                semaphore_state.try_acquire(&mut mut_self.wait_node, cx)
+            };
+
+            match poll_res {
+                Poll::Pending => {
+                    mut_self.semaphore.replace(semaphore);
+                    Poll::Pending
+                }
+                Poll::Ready(()) => {
+                    let to_release = match mut_self.auto_release {
+                        true => mut_self.wait_node.required_permits,
+                        false => 0,
+                    };
+                    Poll::Ready(GenericSharedSemaphoreReleaser::<MutexType> {
+                        semaphore,
+                        permits: to_release,
+                    })
+                }
+            }
+        }
+    }
+
+    impl<MutexType: RawMutex> FusedFuture
+        for GenericSharedSemaphoreAcquireFuture<MutexType>
+    {
+        fn is_terminated(&self) -> bool {
+            self.semaphore.is_none()
+        }
+    }
+
+    impl<MutexType: RawMutex> Drop
+        for GenericSharedSemaphoreAcquireFuture<MutexType>
+    {
+        fn drop(&mut self) {
+            // If this GenericSharedSemaphoreAcquireFuture has been polled and it was added to the
+            // wait queue at the semaphore, it must be removed before dropping.
+            // Otherwise the semaphore would access invalid memory.
+            if let Some(semaphore) = self.semaphore.take() {
+                let mut semaphore_state = semaphore.state.lock();
+                // Analysis: Does the number of permits play a role here?
+                // The future was notified because there was a certain amount of permits
+                // available.
+                // Removing the waiter will wake up as many tasks as there are permits
+                // available inside the Semaphore now. If this is bigger than the
+                // amount of permits required for this task, then additional new
+                // tasks might get woken. However that isn't bad, since
+                // those tasks should get into the wait state anyway.
+                semaphore_state.remove_waiter(&mut self.wait_node);
+            }
+        }
+    }
+
+    /// A futures-aware shared semaphore.
+    pub struct GenericSharedSemaphore<MutexType: RawMutex> {
+        state: Arc<LockApiMutex<MutexType, SemaphoreState>>,
+    }
+
+    impl<MutexType: RawMutex> PartialEq for GenericSharedSemaphore<MutexType> {
```

Comment:
I think the other shared types do not implement `PartialEq`, or do they? In doubt I would rather leave it out, since `PartialEq` in Rust typically seems more about content equality than reference equality.

Reply:
Agreed. An explicit ptr equality method would be much better. Unsure what I was thinking.